File: codechef/MAY16/FORESTGA.py
MAX_T=1000000000000000000L
MAX=100001
n=0
w=0
l=0
h=[]
r=[]

def findNew(t):
	maxredeemption=0
	for i in range(0,n):
		newh=h[i]+r[i]*t
		if newh>=l:
			maxredeemption=maxredeemption+newh
		if maxredeemption>w:
			return maxredeemption
	return maxredeemption;

def findAnswer():
	startt=0
	endt=MAX_T
	if findNew(0)>=w:
		return 0;
	while 1==1:
		t=(startt+endt)/2
		w1=findNew(t)
		w2=findNew(t-1)
		w3=findNew(t+1)
		
		if w2<w and w1>=w:
			return t 
		if w3>=w and w1<w:
			return t+1;
		
		if w1>w: 
			endt=t-1 
		else :
			startt=t+1 
		
		if startt>endt:
			break;
	return t

n,w,l=map(int,raw_input().split())
for i in range(0,n):
	a,b=map(int,raw_input().split())
	h.append(a)
	r.append(b)
print findAnswer()

File: codechef/MAY16/FORESTGA.c
#include<stdio.h>
typedef long long int ll;
#define MAX_T 1000000000
#define MAX 100001
ll n,w,l,h[MAX],r[MAX],newh[MAX];

ll findNew(ll t)
{
	ll maxredeemption=0,i;
	for(i=0;i<n;i++)
	{	
		newh[i]=h[i]+r[i]*t;
		if(newh[i]>=l)
			maxredeemption+=newh[i];
	}
	return maxredeemption;
}
ll findAnswer()
{
	ll w1,w2,w3,startt=0,endt=MAX_T,t;
	if(findNew(0)>=w)
		return 0;
	while(1)
	{
		t=(startt+endt)/2;
		
		w1=findNew(t);
		w2=findNew(t-1);
		w3=findNew(t+1);
		
		//printf("hello lavish kothari %lld %lld %lld\n",t,w1,w2);
		
		if(w2<w && w1>=w)
			return t;
		if(w3>=w && w1<w)
			return t+1;
		if(w1>w)
			endt=t-1;
		else startt=t+1;
		if(startt>=endt)
			break;
	}
	return t;
}
int main()
{
	ll i;
	scanf("%lld%lld%lld",&n,&w,&l);
	for(i=0;i<n;i++)
		scanf("%lld%lld",h+i,r+i);
	printf("%lld\n",findAnswer());
	return 0;
}
